File: top_interview_questions_easy/two_sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
      unordered_map<int, int> map;
      for(int i = 0; i < nums.size(); ++i) {
        if(map.find(target - nums[i]) != map.end()) {
          return {map[target - nums[i]], i};
        }
      map.insert({nums[i], i});
      }
      return {};
    }
};

// unordered_map
// 원소들이 순서대로 정렬되어 들어가지 않음
// 삽입, 삭제, 조회가 평균 O(1)을 보장(최악의 경우 O(N))
// 해시 함수 사용해 저장공간 결정
// 너무 빈번하게 자료 삽입, 삭제하지 않고 많은 자료저장 및 빠른 검색 속도를 원할 때 사용

File: top_interview_questions_easy/two_sum.py

class Solution:
    """
    approach
    큰 for문을 돌면서 한 정수에 대해 다른 값들을 맞추어 가며 합이 target이 되는지 확인한다.

    fail
    i + 1 과 같이 임의로 증가시킨 index 가 range out 이 되어 버린다.

    """
    def twoSum(self, nums: List[int], target: int) -> List[int]:
            for i in range(len(nums)):
                for j in range(i+1, len(nums)):
                    if nums[i] + nums[j] == target:
                          return [i, j]

    """
    success code
    hash_table을 사용한다.
    nums 를 돌면서 nums[i]를 hash_table에 넣는다.
    이후 해당 hash_table에는 nums[i]와 더하면 target이 나오는 값, 즉 target - nums[i]가 들어간다.
    """
    def twoSum(self, nums: List[int], target:int) -> List[int]:
        hash_table = {}
        for i in range(len(nums)):
            if nums[i] in hash_table:
                return [hash_table[nums[i]], i]
            else:
                hash_table[target-nums[i]] = i



